indexing-service/src/main/java/org/apache/druid/indexing/overlord/ZkWorker.java

```diff
@@ -63,6 +67,25 @@ public ZkWorker(Worker worker, PathChildrenCache statusCache, final ObjectMapper
     this.statusCache = statusCache;
     this.cacheConverter = (ChildData input) ->
         JacksonUtils.readValue(jsonMapper, input.getData(), TaskAnnouncement.class);
+    this.taskIdExtractor = createTaskIdExtractor(jsonMapper);
+  }
+
+  static java.util.function.Function<ChildData, String> createTaskIdExtractor(final ObjectMapper jsonMapper)
+  {
+    return (ChildData input) -> {
+      try (JsonParser parser = jsonMapper.getFactory().createParser(input.getData())) {
+        while (parser.nextToken() != JsonToken.END_OBJECT) {
+          if ("id".equals(parser.getCurrentName())) {
```

Comment:
it would probably be worthwhile to put some safeguards to ensure we don't accidentally extract an "id" field from one of the nested objects

Reply:
agreed, good suggestion